### ISSUE TITLE
Enable ECP/Deep Linking

### ISSUE THREAD

**Issue by pixshatterer** [rokucommunity/vscode-brightscript-language#4]
Is there a way to pass arguments to simulate deeplinking while debugging?

**Comment by TwitchBronBron** [comment]
Not currently, no. Would another launch.json setting work for you? The new setting could be called "deepLinkUrl".  Something like this:


```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "brightscript",
            "request": "launch",
            "name": "BrightScript Debug: Launch",
            "host": "192.168.1.17",
            "deepLinkUrl": "http://192.168.1.17:8060/install/<channel_id>?contentID=<contentID>&mediaType=series",
            "password": "password",
            "rootDir": "${workspaceFolder}",
            "consoleOutput": "full",
            "stopOnEntry": false
        }
    ]
}

```

**Comment by pixshatterer** [comment]
that would be awesome! also an unrelated question - the plugin uses a `rokudeploy.json` while deploying? since it uses `roku-deploy`  module, I wondered if that was the case.

**Comment by TwitchBronBron** [comment]
Yes, you should technically be able to use a rokudeploy.json file. I believe the `files` property doesn't work in rokudeploy.json when using this extension, and there may be a few others that get overridden by this extension. 

However, an undocumented alternative to rokudeploy.json is that you can actually specify rokudeploy.json settings directly in your launch configuration. I pass the entire launch configuration object to rokudeploy, and it uses the options it knows about.

**Comment by TwitchBronBron** [comment]
I just wanted to give you an update. I started working on the "launch directly to deep link" concept a while back, and made some progress. Life just got busy. It's possible, and I'll try to get to it sometime soon...but just wanted to give an update saying that work has at least been started.

**Comment by TwitchBronBron** [comment]
@pixshatterer I just pushed [this branch](https://github.com/TwitchBronBron/vscode-brightscript-language/tree/feature/deep-linking) with my first attempt at supporting deep linking. Can you pull down the code and try it out and let me know how it goes?

**Comment by pixshatterer** [comment]
@TwitchBronBron sure! how would be the usage? it is still the `"deepLinkUrl"` property the one to use?

**Comment by TwitchBronBron** [comment]
Yep, that's the one! Use "deepLinkUrl". You can use the ${host} placeholder in that url if you don't want to duplicate the hostname, but I would test it out with a full url at first to make sure it's working.

**Comment by TwitchBronBron** [comment]
@pixshatterer Have you had a chance to look at this change yet? I'd love to get it merged into master, but I want to make sure it works for a proper roku app (not just my test apps) before I do.

**Comment by TwitchBronBron** [comment]
@pixshatterer @georgejecook @triwav @bvisin I just merged master into this branch, so it should have all of the latest changes now. 

Does this work well enough for you all that it can be merged into master / is there anything else that needs done?

**Comment by TwitchBronBron** [comment]
The latest version of the branch has remove `${promptForContentId}` and `${promptForMediaType} in favor of `${promptForQueryParams}`. This should hopefully streamline the process a little bit.

**Comment by timalacey** [comment]
Anything stopping this from being merged into master?

**Comment by TwitchBronBron** [comment]
@timalacey  I don't do any Roku development right now, so I need some actual Roku developers to test it out and validate it before I merge. If you're interested in trying it out, and don't want to figure out how to build the project locally, let me know and I can get you a `vsix` bundle with these changes included for you test out.

**Comment by timalacey** [comment]
Apologies for not getting back to you @TwitchBronBron 

It works. I'm not too concerned with the 1.5 seconds lost due to the auto-run and then reacting to the /launch ECP.
Your feature also works on older Roku's which Eclipse never did, so thanks.

**Comment by pixshatterer** [comment]
Apologies for not getting back to you @TwitchBronBron too!  totally swamped with non-roku projects right now so I had no chance to check this out - I would say go ahead!

**Comment by LasseDenecken** [comment]
@TwitchBronBron I'm not able to get this to work with a **clean vscode installation** and **without using brighterscript**. 
Here is what I did: 
1. Installed vscode
2. Installed the vscode-brightscript-language extension
3. Created the most basic roku app (manifest, source/main.brs, components/AppScene.xml, components/AppScene.brs)
4. Added the default launch.json (and changed the host)
5. Added "deepLinkUrl": "http://${host}:8060/launch/dev?contentId=1&mediaType=episode"
The dev channel is being launched, but the expected relaunch with the deeplink-parameters is not happening.
Did I miss something?

**Comment by TwitchBronBron** [comment]
Hmm. I'm having trouble getting this to work as well. This must have broken at some point. @chrisdp we might want to take this opportunity to revisit the ECP/Deep Linking logic in general. Perhaps inject the values on the first line of the main function?

**Comment by triwav** [comment]
I’m a strong proponent of that approach. The current method requires a restart of the application which is kind of clunky and has issues with the debug protocol as well I believe

**Comment by TwitchBronBron** [comment]
Do we need to worry about things like URL encoding?=

**Comment by chrisdp** [comment]
No. The device runs one decode on the way in so we would inject as typed.

**Comment by chrisdp** [comment]
We should also update it so you can define a deeplink as such:

```
{
MediaType: string | string from enum( will get list),
ContentId: string
}
```

**Comment by chrisdp** [comment]
list of media types from [Roku documentation](https://developer.roku.com/docs/developer-program/discovery/implementing-deep-linking.md#mediatype-behavior):
```
movie
episode
season
series
shortFormVideo
special
live
```

**Comment by chrisdp** [comment]
relates to #192

**Comment by LasseDenecken** [comment]
@chrisdp besides the actual deeplinking, you can also run the [unit-testing-framework](https://github.com/rokudev/unit-testing-framework) via deeplink.
Running tests has been mentioned in this issue: https://github.com/rokucommunity/vscode-brightscript-language/issues/53

**Comment by simon-grantham** [comment]
Actually something like this approach in launch.json would be more useful:
			"args": [ "mediaType=episode", "contentId=some_bunch_of_ids" ],

This way one could also set something like:
			"args": [ "my_Debugging_arg=my_Debugging_val" ]
